rir/src/compiler/opt/type_speculation.cpp

```diff
@@ -16,106 +17,64 @@ void TypeSpeculation::apply(RirCompiler&, ClosureVersion* function,
-                                    seen.isA(PirType(RType::real)
-                                                 .orPromiseWrapped()) ||
-                                    seen.isA(PirType(RType::logical)
-                                                 .orPromiseWrapped())) {
-                                    assume = seen;
-                                }
-                                if (!input->type.isA(assume)) {
-                                    speculate[cp][input] = {
-                                        assume, force->typeFeedback.srcCode,
-                                        force->typeFeedback.origin};
-                                    force->observed =
-                                        Force::ArgumentKind::unknown;
-                                    trigger = false;
-                                }
-                            }
-                        }
+                        speculateOn = arg;
+                        break;
+                    case Force::ArgumentKind::evaluatedPromise:
+                        speculateOn = arg;
+                        feedback.type = feedback.type.orPromiseWrapped();
                         break;
-                    }
-
                     case Force::ArgumentKind::promise:
-                    case Force::ArgumentKind::unknown: {
-                        force->observed = Force::ArgumentKind::unknown;
+                    case Force::ArgumentKind::unknown:
                         break;
                     }
-
-                    default:
-                        assert(false && "unknown observed force input kind");
-                    }
                 }
             }
-            if (trigger) {
-                PirType seen = i->typeFeedback.type;
-                if (!seen.isVoid() && !seen.maybeObj() && !i->type.isA(seen)) {
-                    if (auto cp = checkpoint.next(i)) {
-                        auto assume =
-                            (seen.isA(
-                                 PirType(RType::integer).orPromiseWrapped()) ||
-                             seen.isA(
-                                 PirType(RType::real).orPromiseWrapped()) ||
-                             seen.isA(
-                                 PirType(RType::logical).orPromiseWrapped()))
-                                ? seen
-                                : i->type.notObject();
-                        if (!i->type.isA(assume))
-                            speculate[cp][i] = {assume, i->typeFeedback.srcCode,
-                                                i->typeFeedback.origin};
-                        // Prevent redundant speculation
-                        i->typeFeedback.type = PirType::bottom();
-                    }
-                }
-            }
-            ip = next;
+        }
+
+        if (!speculateOn)
+            return;
+
+        if (auto cp = checkpoint.next(i)) {
```

Comment:
For the case, this used to be `auto cp = checkpoint.at(force)` and not `next`

Reply:
I see. guess `checkpoint.next(speculateOn)` makes the most sense, doesn't it?

Reply:
the reason I am working on this is, that I saw in matrix-for we added a sub-par test and I wanted to add more types.

so I started with this refactoring, that maybe should have been done earlier...

Reply:
Good point.. My reasoning was to not have the checkpoint and the assume too far apart, but I guess this allows to hoist the assume further.

By more types you mean something like `(logical | real)'`?

Reply:
I wonder whether `checkpoint.next(speculateOn)` could have as a consequence that less checkpoints are available